torch/utils/data/datasets/decoder.py

```diff
@@ -0,0 +1,270 @@
+                continue
+
+            if extension[-len(target):] == target:
+                return f(data)
+            return None
+    return g
+
+
+class ImageHandler:
+    """
+    Decode image data using the given `imagespec`.
+    The `imagespec` specifies whether the image is decoded
+    to numpy/torch/pi, decoded to uint8/float, and decoded
+    to l/rgb/rgba:
+
+    - l8: numpy uint8 l
+    - rgb8: numpy uint8 rgb
+    - rgba8: numpy uint8 rgba
+    - l: numpy float l
+    - rgb: numpy float rgb
+    - rgba: numpy float rgba
+    - torchl8: torch uint8 l
+    - torchrgb8: torch uint8 rgb
+    - torchrgba8: torch uint8 rgba
+    - torchl: torch float l
+    - torchrgb: torch float rgb
+    - torch: torch float rgb
+    - torchrgba: torch float rgba
+    - pill: pil None l
+    - pil: pil None rgb
+    - pilrgb: pil None rgb
+    - pilrgba: pil None rgba
+    """
+    def __init__(self, imagespec):
+        assert imagespec in list(imagespecs.keys()), "unknown image specification: {}".format(imagespec)
+        self.imagespec = imagespec.lower()
+
+    def __call__(self, key, data):
+        extension = re.sub(r".*[.]", "", key)
+        if extension.lower() not in "jpg jpeg png ppm pgm pbm pnm".split():
+            return None
+
+        imagespec = self.imagespec
+        atype, etype, mode = imagespecs[imagespec]
+
+        with io.BytesIO(data) as stream:
+            img = PIL.Image.open(stream)
+            img.load()
+            img = img.convert(mode.upper())
+            if atype == "pil":
+                return img
+            elif atype == "numpy":
+                result = np.asarray(img)
+                assert result.dtype == np.uint8, "numpy image array should be type uint8, but got {}".format(result.dtype)
+                if etype == "uint8":
+                    return result
+                else:
+                    return result.astype("f") / 255.0
+            elif atype == "torch":
+                import torch
+
+                result = np.asarray(img)
+                assert result.dtype == np.uint8, "numpy image array should be type uint8, but got {}".format(result.dtype)
+
+                if etype == "uint8":
+                    result = np.array(result.transpose(2, 0, 1))
+                    return torch.tensor(result)
+                else:
+                    result = np.array(result.transpose(2, 0, 1))
+                    return torch.tensor(result) / 255.0
+            return None
+
+def imagehandler(imagespec):
+    return ImageHandler(imagespec)
+
+
+################################################################
+# torch video
+################################################################
+
+
+def torch_video(key, data):
+    extension = re.sub(r".*[.]", "", key)
+    if extension not in "mp4 ogv mjpeg avi mov h264 mpg webm wmv".split():
+        return None
+
+    # add `type: ignore` to avoid mypy's warning on import missing
+    import torchvision.io  # type: ignore
```

Comment:
Cycle dependency here.

Reply:
why this is cycle?

Reply:
torchvision and torchaudio has dependency on torch, right?

Reply:
Yes, but not the other direction. 
torchvision.io/torchaudio import torch, but torch won't import torchvision.io/torchaudio, there is no cycle here to me, correct me if I am wrong. These code are from Nvidia, I might missed something here.

Reply:
You are right, it's not circular dependency. Great to understand how python executes the import.
But, we should do `try import` here, since we don't explicitly have torchvision or torchaudio. I guess the webdataset has explicit requirements for these packages. (they remove it on October [link](https://github.com/tmbdev/webdataset/commit/f8cf56273d3324b86b22b24e47b2c7d72038de32#diff-4d7c51b1efe9043e44439a949dfd92e5827321b34082903477fd04876edb7552))
Besides, it's wield for me to import a package one time for every single data. For a huge dataset, the overhead can be crazy.

Reply:
Agree, this is not good generally. I think what Nvidia want to achieve here is that `if people don't need this handler, then don't import the needed library`, we can discuss further. I have an impression that python import is something like it check the module dict first, if no such module, then import it, the overhead here should be dict lookup? I am not 100% sure though. Let's revisit this after the tar dataset PR is landed.